### DISCUSSION TITLE
Support for 2FA-Auth

### DISCUSSION THREAD

**Discussion by MDXDave** [craftcms/cms#8453] [Ideas] [answered]
How about 2FA support on login page? (Yubico, Google authenticator etc.)

**Comment by angrybrad** [comment]
Closing this as a dupe of https://github.com/craftcms/cms/issues/1471. Feel free to add additional thoughts/comments over there.

**Comment by brandonkelly** [comment]
Could be argued that 2FA should be an option in core, without requiring a plugin that provides a custom login form.

**Comment by MDXDave** [comment]
Yes, the intention was, to support 2FA on login without any external plugin.

**Comment by roelvanhintum** [comment]
Right now i'm refactoring our 2FA plugin for Craft 3 (it was written for Craft 2). I'll give an update when it's done. https://github.com/born05/craft-twofactorauthentication

**Comment by roelvanhintum** [comment]
@MDXDave the Craft 3 version of our 2FA plugin is in beta.
https://github.com/born05/craft-twofactorauthentication

**Comment by XhmikosR** [comment]
I'd also like to see this in core. 2FA is a mandatory precaution these days :)

**Comment by stevenjwright** [comment]
@brandonkelly Is there any movement on this? Just wondering the likelihood of it ending up being pushed back again? Cheers!

**Comment by brandonkelly** [comment]
@stevenjwright I can’t promise exactly when we’ll get to it, but we have our own internal needs for it so it’ll happen sooner than later.

**Comment by zavelin** [comment]
Hello there - any update on when this will be added? will it be in 4.0?

**Comment by brandonkelly** [comment]
That’s the current plan.

**Comment by roelvanhintum** [comment]
@brandonkelly i'm both happy and sad to read this. ;)

**Comment by Mosnar** [comment]
Since this is on the menu right now, I wanted to request that the two-factor methods supported include TOTP (e.g. Authy and Google Authenticator) and FIDO U2F (e.g. [Yubikey](https://www.yubico.com/), [Titan](https://store.google.com/us/product/titan_security_key?hl=en-US))

Related: [Browser support for FIDO U2F API](https://caniuse.com/u2f)

**Comment by andris-sevcenko** [comment]
@Mosnar Not to re-iterate, but that is the current plan :)

**Comment by admanami** [comment]
Does this mean 2FA is no longer on the roadmap for 4.0? Or that it has been completed?

**Reply by andris-sevcenko** [reply]
It's currently in an MVP state on a branch and will definitely be a part of 4.0

**Reply by admanami** [reply]
I can't see this in the 4.0 beta release notes. I guess this means 2FA did not make it into 4.0?

**Reply by andris-sevcenko** [reply]
@admanami sorry, but no. It's still in an almost finished stage, though.

The reason it got bumped to 4.1 was to have the chance of wrapping up the filesystem and image transformer changes. The latter changes are breaking changes for plugins, while the auth changes are not. So, the filesystem and the like changes _had_ to be part of 4.0, while auth _could_ also be part of 4.1.

In that situation, we had to prioritise the asset stuff over auth stuff.

**Comment by zizther** [comment]
Is this feature still scheduled for a 4.x release. It is not in the roadmap at all, just interested to know

**Reply by brandonkelly** [reply]
We’ve gone back to the drawing board with it, so can’t comment on when exactly it’ll be released.

**Comment by Mosnar** [comment]
I just wanted to check in on this feature and see if it can make it back onto the roadmap, as it's a serious show-stopper for some clients (especially those with deeper pockets).

**Reply by brandonkelly** [reply]
It’s actively being worked on!

**Reply by admanami** [reply]
Personally I feel this should be prioritised as it is pretty essential. After several years, we could not wait any longer and went with Cloudflare ZT Access instead, but it would be good to see this in the core product.

**Reply by brandonkelly** [reply]
I’m not sure how much more we can prioritize something we’re already working on :)

**Reply by admanami** [reply]
I meant earlier. But appreciate your work, fantastic product.

**Reply by plcdnl** [reply]
how is possible to enable it?

**Reply by thupsi** [reply]
@plcdnl It's not possible yet. In the works for Craft 5.

**Comment by aeu** [comment]
I just want to add an observation that may hopefully raise awareness of the urgency of this subject. Our General liability insurance renewed, and in the questionnaire they specifically asked if the software we provide for our clients supported two factor authentication.

I see that this thread was started on August 2017, please get this going soon, and for Craft 4 as well.

**Reply by Anubarak** [reply]
It in the  ~~"recently shipped"~~ edit: "in progress" list for Craft 5
https://craftcms.com/roadmap

However: if it is so important for you, you could implement it on your own. 
We did this for several projects via custom Plugin

**Reply by thupsi** [reply]
Small correction, it's the "in progress" list.

**Reply by Anubarak** [reply]
Oh sorry, stupid me. Thank you for the correction - not sure what was wrong with me 🤣 
But at least we have progress.
I corrected my answer

**Comment by brandonkelly** [comment, marked as answer]
Craft 5 is out now with built-in MFA support 🚀
